src/osd/pg_scrubber.h

```diff
@@ -0,0 +1,677 @@
+   * 'false' means we are configured with no replicas, and
+   * should expect no maps to arrive.
+   */
+  bool get_replicas_maps(bool replica_can_preempt) final;
+
+  Scrub::FsmNext on_digest_updates() final;
+
+  void send_replica_map(bool was_preempted) final;
+
+  void send_remotes_reserved() final;
+  void send_reservation_failure() final;
+
+  /**
+   *  does the PG have newer updates than what we (the scrubber) know?
+   */
+  [[nodiscard]] bool has_pg_marked_new_updates() const final;
+
+  void set_subset_last_update(eversion_t e) final;
+
+  void replica_update_start_epoch() final;
+
+  void maps_compare_n_cleanup() final;
+
+  Scrub::preemption_t* get_preemptor() final;
+
+  int build_primary_map_chunk() final;
+
+  int build_replica_map_chunk() final;
+
+  void reserve_replicas() final;
+
+  [[nodiscard]] bool was_epoch_changed() const final;
+
+  void mark_local_map_ready() final;
+
+  [[nodiscard]] bool are_all_maps_available() const final;
+
+  std::string dump_awaited_maps() const final;
+
+ protected:
+  bool state_test(uint64_t m) const { return m_pg->state_test(m); }
+  void state_set(uint64_t m) { m_pg->state_set(m); }
+  void state_clear(uint64_t m) { m_pg->state_clear(m); }
+
+  [[nodiscard]] bool is_primary() const { return m_pg->recovery_state.is_primary(); }
+
+  [[nodiscard]] bool is_scrub_registered() const;
+
+  virtual void _scrub_clear_state() {}
+
+  utime_t m_scrub_reg_stamp;  ///< stamp we registered for
+
+  ostream& show(ostream& out) const override;
+
+ public:
+  // -------------------------------------------------------------------------------------------
+
+  friend ostream& operator<<(ostream& out, const PgScrubber& scrubber);
+
+  static utime_t scrub_must_stamp() { return utime_t(1, 1); }
+
+  virtual ~PgScrubber();  // must be defined separately, in the .cc file
+
+  [[nodiscard]] bool is_scrub_active() const final;
+
+ private:
+  void reset_internal_state();
+
+  void _scan_snaps(ScrubMap& smap);  // note that the (non-standard for a
+				     // non-virtual) name of the function is searched
+				     // for by the QA standalone tests. Do not modify.
+
+  void clean_meta_map(ScrubMap& for_meta_scrub);
+
+  void run_callbacks();
+
+  /**
+   * are we still a clean & healthy scrubbing primary?
+   *
+   * relevant only after the initial sched_scrub
+   */
+  [[nodiscard]] bool is_event_relevant(epoch_t queued) const;
+
+  /**
+   * check the 'no scrub' configuration options.
+   */
+  [[nodiscard]] bool should_abort_scrub(epoch_t queued) const;
+
+  void send_epoch_changed();
+
+  /**
+   * return true if any inconsistency/missing is repaired, false otherwise
+   */
+  [[nodiscard]] bool scrub_process_inconsistent();
+
+  bool m_needs_sleep{true};  ///< should we sleep before being rescheduled? always
+			     ///< 'true', unless we just got out of a sleep period
+
+
+  // 'optional', as 'ReplicaReservations' & 'LocalReservation' are 'RAII-designed'
+  // to guarantee un-reserving when deleted.
```

Comment:
As I mentioned elsewhere, however, any remote reservations we hold when we get an OSDMap informing us of a new interval will be released by those osds on their own when the get the messages.

For m_[local|remote]_osd_resource, this makes sense as we *must* release the reservation regardless of how we stop scrubbing.  However, wouldn't it be easier if these were embedded in the statemachine state object corresponding to the outermost state in which we hold them?  Otherwise, we don't really benefit from the RAII structure as you need to explicitly call reset().